[AArch64][SVE] Add lowering for PARTIAL_REDUCE_U/SMLA to USDOT

NickGuy-Arm opened the pull request:
Add lowering for PARTIAL_REDUCE_U/SMLA nodes to USDOT instructions. This happens when there is a MUL instruction as the second operand in the ISD node. Then the extends on the operands of the MUL op need to have a different signedness.

@JamesChesterman is the original author

MacDue commented on llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.cpp:
```diff
@@ -924,8 +924,19 @@ SDValue DAGTypeLegalizer::CreateStackStoreLoad(SDValue Op,
   // See if the target wants to custom lower this node.
-  if (TLI.getOperationAction(N->getOpcode(), VT) != TargetLowering::Custom)
-    return false;
+  unsigned Opcode = N->getOpcode();
+  bool IsPRMLAOpcode =
+      Opcode == ISD::PARTIAL_REDUCE_UMLA || Opcode == ISD::PARTIAL_REDUCE_SMLA;
+
+  if (IsPRMLAOpcode) {
+    if (TLI.getPartialReduceMLAAction(N->getValueType(0),
+                                      N->getOperand(1).getValueType()) !=
+        TargetLowering::Custom)
+      return false;
+  } else {
+    if (TLI.getOperationAction(Opcode, VT) != TargetLowering::Custom)
+      return false;
+  }
```
Can't `LegalizeVectorOps` handle this? `getPartialReduceMLAAction()` is already hooked up there and should be able to call into the custom lowering?

MacDue replied:
Is this done to bypass type legalization for the `usdot_8to64` case? Could we handle that instead by adding a combine that reduces accumulators of `<vscale x 4 x i64>` to `<vscale x 4 x i32>` followed by a extend?

NickGuy-Arm replied:
It isn't strictly to work around type legalization, but to avoid the extended type being split before we can account for it properly. The type coming in via `VT` is the type of the operand of the `partial_reduce_umla` node, which is an extend, so it effectively hides the actual operand type at this stage.
I don't think the pre-legalization DAG combine would work for the reasons you pointed out, but in trying to implement the separate node, I encountered the exact same issues.
I've added an operation action for `ISD::PARTIAL_REDUCE_UMLA` with `nxv16i32`, which is the post-extended type of `nxv16i8`, and we can have the existing validation within `LowerPARTIAL_REDUCE_MLAToUSDOT` decide whether it can actually be lowered to USDOT (falling back to unpacks and `mla`s if not).

NickGuy-Arm replied:
I've reimplemented this check, as I believe it is the simplest solution to this problem. For `USDOT` lowering to function, it needs to happen pre-legalization because it deals with illegal intermediate types (which are then flattened out by replacing the nodes with the `USDOT` ISD node). 
As the partialReduceMLA LegalizeActions are handled differently from the standard operation actions, we need to check the relevant action to take.

This check is simply the required plumbing to have the legalizer respect when a target says that it has custom lowering for a given partial reduction. If we try to pack the information into the operation actions, we lose the ability to filter based on what the partial reduction is reducing from. And trying to move the check to post-legalization we lose direct access to the pre-extend type, as the nodes required to legalize the type obscure it through multiple extends or AArch64ISD unpack nodes.

NickGuy-Arm committed:
[AArch64][SVE] Add lowering for PARTIAL_REDUCE_U/SMLA to USDOT

Add lowering for PARTIAL_REDUCE_U/SMLA nodes to USDOT instructions.
This happens when there is a MUL instruction as the second operand
in the ISD node. Then the extends on the operands of the MUL op
need to have a different signedness.

NickGuy-Arm committed:
Add calls to setPartialReduceMLAAction.

NickGuy-Arm committed:
Rebase and update tests

NickGuy-Arm committed:
Adjust how usdot cases are lowered

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -7742,8 +7755,8 @@ SDValue AArch64TargetLowering::LowerOperation(SDValue Op,
     return LowerVECTOR_HISTOGRAM(Op, DAG);
-  case ISD::PARTIAL_REDUCE_SMLA:
   case ISD::PARTIAL_REDUCE_UMLA:
+  case ISD::PARTIAL_REDUCE_SMLA:
```
nit: unnecessary change.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1869,7 +1869,20 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
+      setPartialReduceMLAAction(MVT::nxv8i16, MVT::nxv16i16, Custom);
+      setPartialReduceMLAAction(MVT::nxv16i8, MVT::nxv32i8, Custom);
+
+      setOperationAction(ISD::PARTIAL_REDUCE_UMLA, MVT::nxv16i32, Custom);
```
This looks like you're trying to work around a problem, because it's now using two separate mechanisms to determine whether a PARTIAL_REDUCE_* node needs custom lowering.

What issues are you running into when implementing this as a DAGcombine that runs before type legalisation?

NickGuy-Arm replied:
The issue presents itself as the extend type being obscured by numerous (depending on the illegal types) pairs of AArch64ISD::UUNPKLO/HI. While it would technically be possible to traverse these to find the pre-extended type, we would have to traverse the entire tree of a potentially unknown size to verify that it's all valid.
By running this lowering before type legalization, we can make our adjustments and insert the node hiererchy representing USDOT, which itself can then benefit from type legalization.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1869,7 +1869,20 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
+
+    if (Subtarget->hasMatMulInt8()) {
+      // USDOT
+      setPartialReduceMLAAction(MVT::nxv2i64, MVT::nxv8i64, Custom);
+      setPartialReduceMLAAction(MVT::nxv4i32, MVT::nxv16i32, Custom);
+      setPartialReduceMLAAction(MVT::nxv2i64, MVT::nxv4i64, Custom);
+      setPartialReduceMLAAction(MVT::nxv4i32, MVT::nxv8i32, Custom);
+      setPartialReduceMLAAction(MVT::nxv8i16, MVT::nxv16i16, Custom);
+      setPartialReduceMLAAction(MVT::nxv16i8, MVT::nxv32i8, Custom);
```
I'd expect the cases you've handled above (where all types are legal) to be sufficient. Why did you have to add this?

NickGuy-Arm replied:
These include the "illegal but internally legalizable" types that were needed for usdot_8to64 cases, these have now been removed from this PR along with the support for 8to64 cases.

NickGuy-Arm committed:
Adjust how usdot cases are lowered

NickGuy-Arm commented:
Converting this to a draft, as an acceptable solution will likely require more effort than I can afford right now. I'll revisit it in the future.
